src/charm.py

```diff
@@ -1,104 +1,377 @@
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
         self.unit.status = ActiveStatus()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+    def _on_start(self, event: StartEvent):
+        """Triggered when on start. Set the right node role."""
+        if (
+            self.opensearch.is_started()
+            and self.app_peers_data.get("security_index_initialised") is not None
+        ):
+            return
+
+        if not self._is_tls_fully_configured():
+            event.defer()
+            return
+
+        self.opensearch_config.set_client_auth()
+
+        try:
+            nodes = self._get_nodes()
+        except OpenSearchHttpError:
+            event.defer()
+            return
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
+        # Set the configuration of the node
+        self._set_node_conf(nodes)
 
-        Learn more about config at https://juju.is/docs/sdk/config
+        # Remove some config entries when cluster bootstrapped
+        self._cleanup_conf_if_bootstrapped(nodes)
+
+        # start opensearch
+        if not self._start_opensearch():
+            event.defer()
+            return
+
+        # initialize the security index if the admin certs are written on disk
+        if self.unit.is_leader() and self.app_peers_data.get("security_index_initialised") is None:
+            admin_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
+            self._initialize_security_index(admin_secrets)
+            self.app_peers_data["security_index_initialised"] = "True"
+
+    def _on_peer_relation_joined(self, _: RelationJoinedEvent):
+        """New node joining the cluster."""
+        current_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
+
+        # In the case of the first unit
+        if current_secrets is None:
+            return
+
+        # Store the "Admin" certificate, key and CA on the disk of the new unit
+        self._store_tls_resources(CertType.APP_ADMIN, current_secrets, override_admin=False)
+
+    def _on_peer_relation_changed(self, event: RelationChangedEvent):
+        """Restart node when cert renewal for the transport layer."""
+        if self.unit_peers_data.get("must_reboot_node") == "True":
+            try:
+                self.opensearch.restart()
+                del self.unit_peers_data["must_reboot_node"]
+            except OpenSearchStartError:
+                event.defer()
+
+    def _on_update_status(self, _: UpdateStatusEvent):
+        """On update status event.
+
+        We want to periodically (every 6 hours) check if certs are expiring soon (in 24h),
```

Comment:
This timing might need adjusting - if it occurs over the weekend, for example, we're out of luck. Maybe up the expiry check to a week in advance?